File: virtual_keyboard.cpp
// virtual_keyboard.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <cstdlib>

using namespace std;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string BLUE = "\033[94m";
const string YELLOW = "\033[93m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

vector<vector<string>> KEYBOARD_EN = {
    {"1","2","3","4","5","6","7","8","9","0"},
    {"Q","W","E","R","T","Y","U","I","O","P"},
    {"A","S","D","F","G","H","J","K","L"},
    {"Z","X","C","V","B","N","M"}
};

vector<vector<string>> KEYBOARD_RU = {
    {"1","2","3","4","5","6","7","8","9","0"},
    {"Й","Ц","У","К","Е","Н","Г","Ш","Щ","З"},
    {"Ф","Ы","В","А","П","Р","О","Л","Д"},
    {"Я","Ч","С","М","И","Т","Ь"}
};

class VirtualKeyboard {
public:
    string lang;
    string text;
    bool shift;
    vector<vector<string>> layout;

    VirtualKeyboard(string l = "en") : lang(l), shift(false) {
        layout = (lang == "en") ? KEYBOARD_EN : KEYBOARD_RU;
    }

    void display() {
        system("clear");
        cout << colorize("🖥️  ВИРТУАЛЬНАЯ КЛАВИАТУРА", BOLD) << endl;
        cout << colorize("Управление: номер клавиши → добавить, 0 → пробел, - → удалить, . → выход", YELLOW) << endl;
        cout << colorize("8 → переключить раскладку, 9 → переключить регистр (Shift)\n", YELLOW) << endl;
        cout << colorize("Текст: " + text, GREEN) << endl << endl;

        int idx = 1;
        for (auto& row : layout) {
            string line;
            for (auto& ch : row) {
                string displayCh = shift ? ch : ch;
                // для нижнего регистра в C++ неудобно, но для демонстрации оставим как есть
                line += colorize(to_string(idx).substr(0,2), BLUE) + ":" + colorize(displayCh, BLUE) + "  ";
                idx++;
            }
            cout << line << endl;
        }
        cout << "\nСпециальные: 0-пробел, --удалить, .-выход, 8-язык, 9-Shift" << endl;
    }

    void run() {
        string cmd;
        while (true) {
            display();
            cout << colorize("Введите номер команды: ", BOLD);
            cin >> cmd;
            if (cmd == ".") break;
            if (cmd == "-") {
                if (!text.empty()) text.pop_back();
                continue;
            }
            if (cmd == "0") {
                text += ' ';
                continue;
            }
            if (cmd == "8") {
                lang = (lang == "en") ? "ru" : "en";
                layout = (lang == "en") ? KEYBOARD_EN : KEYBOARD_RU;
                continue;
            }
            if (cmd == "9") {
                shift = !shift;
                continue;
            }
            try {
                int num = stoi(cmd);
                vector<string> flat;
                for (auto& row : layout)
                    for (auto& ch : row)
                        flat.push_back(ch);
                if (num >= 1 && num <= (int)flat.size()) {
                    string ch = flat[num-1];
                    if (shift) {
                        // преобразовать в верхний регистр (только для латиницы)
                        for (char& c : ch) c = toupper(c);
                    } else {
                        for (char& c : ch) c = tolower(c);
                    }
                    text += ch;
                } else {
                    cout << colorize("Неверный номер!", RED) << endl;
                    cin.ignore();
                    cin.get();
                }
            } catch (...) {
                cout << colorize("Неверный ввод!", RED) << endl;
                cin.ignore();
                cin.get();
            }
        }
        cout << colorize("\nИтоговый текст: " + text, GREEN) << endl;
    }
};

int main() {
    VirtualKeyboard kb("en");
    kb.run();
    return 0;
}

File: virtual_keyboard.cs
// virtual_keyboard.cs
using System;
using System.Collections.Generic;
using System.Linq;

class VirtualKeyboard
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "blue" => "\x1b[94m",
            "yellow" => "\x1b[93m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static List<List<string>> KEYBOARD_EN = new List<List<string>>
    {
        new List<string>{"1","2","3","4","5","6","7","8","9","0"},
        new List<string>{"Q","W","E","R","T","Y","U","I","O","P"},
        new List<string>{"A","S","D","F","G","H","J","K","L"},
        new List<string>{"Z","X","C","V","B","N","M"}
    };

    static List<List<string>> KEYBOARD_RU = new List<List<string>>
    {
        new List<string>{"1","2","3","4","5","6","7","8","9","0"},
        new List<string>{"Й","Ц","У","К","Е","Н","Г","Ш","Щ","З"},
        new List<string>{"Ф","Ы","В","А","П","Р","О","Л","Д"},
        new List<string>{"Я","Ч","С","М","И","Т","Ь"}
    };

    private string lang;
    private List<string> text = new List<string>();
    private bool shift;
    private List<List<string>> layout;

    public VirtualKeyboard(string lang = "en")
    {
        this.lang = lang;
        this.shift = false;
        this.layout = lang == "en" ? KEYBOARD_EN : KEYBOARD_RU;
    }

    public void Display()
    {
        Console.Clear();
        Console.WriteLine(Colorize("🖥️  ВИРТУАЛЬНАЯ КЛАВИАТУРА", "bold"));
        Console.WriteLine(Colorize("Управление: номер клавиши → добавить, 0 → пробел, - → удалить, . → выход", "yellow"));
        Console.WriteLine(Colorize("8 → переключить раскладку, 9 → переключить регистр (Shift)\n", "yellow"));
        Console.WriteLine(Colorize("Текст: " + string.Join("", text), "green"));
        Console.WriteLine();

        int idx = 1;
        foreach (var row in layout)
        {
            string line = "";
            foreach (string ch in row)
            {
                string displayCh = shift ? ch.ToUpper() : ch.ToLower();
                line += Colorize(idx.ToString("D2"), "blue") + ":" + Colorize(displayCh, "blue") + "  ";
                idx++;
            }
            Console.WriteLine(line);
        }
        Console.WriteLine("\nСпециальные: 0-пробел, --удалить, .-выход, 8-язык, 9-Shift");
    }

    public void Run()
    {
        while (true)
        {
            Display();
            Console.Write(Colorize("Введите номер команды: ", "bold"));
            string cmd = Console.ReadLine().Trim();
            if (cmd == ".")
                break;
            if (cmd == "-")
            {
                if (text.Count > 0) text.RemoveAt(text.Count - 1);
                continue;
            }
            if (cmd == "0")
            {
                text.Add(" ");
                continue;
            }
            if (cmd == "8")
            {
                lang = lang == "en" ? "ru" : "en";
                layout = lang == "en" ? KEYBOARD_EN : KEYBOARD_RU;
                continue;
            }
            if (cmd == "9")
            {
                shift = !shift;
                continue;
            }
            if (int.TryParse(cmd, out int num))
            {
                var flat = layout.SelectMany(row => row).ToList();
                if (num >= 1 && num <= flat.Count)
                {
                    string ch = flat[num - 1];
                    ch = shift ? ch.ToUpper() : ch.ToLower();
                    text.Add(ch);
                }
                else Console.WriteLine(Colorize("Неверный номер!", "red"));
            }
            else Console.WriteLine(Colorize("Неверный ввод!", "red"));
        }
        Console.WriteLine(Colorize("\nИтоговый текст: " + string.Join("", text), "green"));
    }

    static void Main()
    {
        var kb = new VirtualKeyboard("en");
        kb.Run();
    }
}
